components/raftstore/src/store/peer.rs

```diff
@@ -67,36 +65,45 @@ pub enum StaleState {
     LeaderMissing,
 }
 
-/// Meta information about proposals.
-pub struct ProposalMeta {
-    pub index: u64,
-    pub term: u64,
-    /// `renew_lease_time` contains the last time when a peer starts to renew lease.
-    pub renew_lease_time: Option<Timespec>,
+struct ProposalQueue<E>
+where
+    E: KvEngine,
+{
+    queue: VecDeque<Proposal<E>>,
 }
 
-#[derive(Default)]
-struct ProposalQueue {
-    queue: VecDeque<ProposalMeta>,
-}
+impl<E: KvEngine> ProposalQueue<E> {
+    fn new() -> ProposalQueue<E> {
+        ProposalQueue {
+            queue: VecDeque::new(),
+        }
+    }
 
-impl ProposalQueue {
-    fn pop(&mut self, term: u64) -> Option<ProposalMeta> {
-        self.queue.pop_front().and_then(|meta| {
-            if meta.term > term {
-                self.queue.push_front(meta);
-                return None;
+    fn find_propose_time(&self, index: u64, term: u64) -> Option<Timespec> {
+        for p in self.queue.iter() {
+            if p.index == index && p.term == term {
+                return Some(p.renew_lease_time.unwrap());
             }
-            Some(meta)
-        })
+        }
+        None
     }
 
-    fn push(&mut self, meta: ProposalMeta) {
-        self.queue.push_back(meta);
+    fn take(&mut self, index: u64, term: u64) -> Vec<Proposal<E>> {
+        let mut propos = Vec::new();
+        while let Some(p) = self.queue.pop_front() {
+            if (p.term, p.index) > (term, index) {
```

Comment:
It is not obvious, which one is significant? `term` or `index`? Please add some comments or rewrite it into two `if` comparison.